src/storage/core.rs

```diff
@@ -649,9 +650,12 @@ where
                 ErrorKind::Bincode(_) => true,
                 ErrorKind::Validation { kind, cause: _ } => {
                     !matches!(kind, ValidationErrorKind::BlobVersion)
-                }
+                },
                 _ => false,
             };
+        } else if let Some(bincode::ErrorKind::Io(error)) =
+            error.downcast_ref::<bincode::Error>().and_then(|b| Some(b.as_ref())) {
+            return error.kind() == std::io::ErrorKind::UnexpectedEof;
```

Comment:
You are trying to detect the error that is not correspond to the current level of abstraction. From the arhitectural point of view it is much better to detect this error at the place, where it happened and convert it to another error from our own `ErrorKind` (it seems that the error can be converted to `ErrorKind::Bincode`)